REGRESSION: Thumbnail turns blank after choose A3/Tabloid paper size in Print Screen

rr-codes opened the pull request:
#### 0ad581ba4f4a68ed4ccdad22996301ffde891b2f
<pre>
REGRESSION: Thumbnail turns blank after choose A3/Tabloid paper size in Print Screen
<a href="https://bugs.webkit.org/show_bug.cgi?id=255210">https://bugs.webkit.org/show_bug.cgi?id=255210</a>
rdar://107503946

Reviewed by Wenson Hsieh.

iOS print preview thumbnail generation uses `CGBitmapContextCreateImage` to create the bitmap image.
However, with some combinations of page scale size with paper size, the resulting image dimension
becomes too large, resulting in the function returning null and no thumbnail being generated.

This PR fixes this by scaling down the dimensions before the image is created to ensure the maximum
possible size of the image is constrained. This limit is the size at which the thumbnail itself is
rendered, so the scaling is safe to do.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToImage):

Canonical link: <a href="https://commits.webkit.org/262780@main">https://commits.webkit.org/262780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4cc22f2dcdd649728c58617e924be72849b9927

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2582 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2586 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2977 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2673 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2282 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2606 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2313 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3749 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/551 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2298 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2161 "149 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2640 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2322 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3515 "264 api tests failed or timed out") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2348 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2131 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2301 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/636 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2337 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2487 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -1042,6 +1042,11 @@ - (void)_createImage:(_WKPrintFormattingAttributes *)formatterAttributes printFo
+                [printFormatter _setPrintPreviewImage:nullptr];
+                return;
+            }
+
```
It's a bit unclear to me why we need this early return.

rr-codes replied:
Because technically, both the retain ptr and the actual ptr can both be nil. If the former, this method would’ve previously actually crashed while attempting to access ‘image.get()’. I know the latter check isn’t strictly necessary, but figured it helps for readability.

whsieh replied:
I'm still a bit confused — it's safe to access `.get()` on a `RetainPtr<CGImageRef>` — you'll get `nil`. So I think this logic boils down to:

```
if image is null:
    set the preview image to null
else:
    set the preview image to image
```

...which seems to be equivalent to just:

```
set the preview image to image
```

rr-codes replied:
ahh I see; I didn't know it was safe to do that. will fix

rr-codes committed:
REGRESSION: Thumbnail turns blank after choose A3/Tabloid paper size in Print Screen
https://bugs.webkit.org/show_bug.cgi?id=255210
rdar://107503946

Reviewed by Wenson Hsieh.

iOS print preview thumbnail generation uses `CGBitmapContextCreateImage` to create the bitmap image.
However, with some combinations of page scale size with paper size, the resulting image dimension
becomes too large, resulting in the function returning null and no thumbnail being generated.

This PR fixes this by scaling down the dimensions before the image is created to ensure the maximum
possible size of the image is constrained. This limit is the size at which the thumbnail itself is
rendered, so the scaling is safe to do.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::drawToImage):

Canonical link: https://commits.webkit.org/262780@main
